src/document_basic.c

```diff
@@ -108,7 +111,70 @@ void Document_MakeRefOwner(Document *doc) {
   doc->flags |= DOCUMENT_F_OWNREFS;
 }
 
+static inline timespec timespecFromMilliseconds(int64_t totalMilliseconds) {
+  timespec result = {.tv_sec = 0, .tv_nsec = 0};
+  if (totalMilliseconds > 0) {
+    result.tv_sec = totalMilliseconds / MILLISECOND_IN_ONE_SECOND;
+    result.tv_nsec = (totalMilliseconds % MILLISECOND_IN_ONE_SECOND) * NANOSECOND_IN_ONE_MILLISECOND;
+  }
+  return result;
+}
+
+
+static inline FieldExpiration* callHashFieldExpirationTime(RedisModuleCtx* ctx, RedisModuleString *keystr, RedisModuleString** fields) {
+  size_t vectorElementCount = array_len(fields);
+  long long fieldCount = vectorElementCount;
+  RedisModuleCallReply *rep = RedisModule_Call(ctx, "HPEXPIRETIME", "sclv", keystr, "FIELDS", fieldCount, fields, vectorElementCount);
+  if (rep == NULL) {
+    RedisModule_Log(ctx, "warning", "Error calling HPEXPIRETIME, error: %d", errno);
+    return NULL;
+  }
+  RS_LOG_ASSERT_FMT(RedisModule_CallReplyType(rep) == REDISMODULE_REPLY_ARRAY, "HPEXPIRETIME returned unexpected reply type: %d", RedisModule_CallReplyType(rep));
+  size_t sz = RedisModule_CallReplyLength(rep);
+  arrayof(FieldExpiration) result = NULL;
+  for (t_fieldIndex i = 0; i < sz; i++) {
+    const int64_t milliseconds = RedisModule_CallReplyInteger(RedisModule_CallReplyArrayElement(rep, i));
+    if (milliseconds < 0) {
+      continue;
+    }
+    const FieldExpiration fieldExpiration = { .index = i, .point = timespecFromMilliseconds(milliseconds)};
+    array_ensure_append_1(result, fieldExpiration);
+  }
+  RedisModule_FreeCallReply(rep);
+  return result;
+}
+
+static inline FieldExpiration* getHashFieldExpirationTime(const IndexSpec *spec, RedisModuleCtx *ctx, RedisModuleString *key) {
+  arrayof(RedisModuleString *) fields = array_newlen(RedisModuleString *, spec->numFields);
+  for (t_fieldIndex field = 0; field < spec->numFields; ++field) {
+    const FieldSpec *f = spec->fields + field;
+    fields[f->index] = RedisModule_CreateString(ctx, f->name, strlen(f->name));
+  }
+  FieldExpiration* result = callHashFieldExpirationTime(ctx, key, fields);
+  for (t_fieldIndex field = 0; field < spec->numFields; ++field) {
+    RedisModule_FreeString(ctx, fields[field]);
+  }
+  array_free(fields);
+  return result;
+}
+
+static inline t_expirationTimePoint getDocExpirationTime(RedisModuleCtx* ctx, RedisModuleKey *openedKey) {
+  t_expirationTimePoint zero = {.tv_sec = 0, .tv_nsec = 0};
+  mstime_t totalMilliseconds = RedisModule_GetAbsExpire(openedKey);
+  if (totalMilliseconds == REDISMODULE_NO_EXPIRE) {
+    return zero;
+  }
+
+  t_expirationTimePoint result = timespecFromMilliseconds(totalMilliseconds);
+  return result;
+}
+
 int Document_LoadSchemaFieldHash(Document *doc, RedisSearchCtx *sctx, QueryError *status) {
+  // must happen before opening the key, in case the call will cause a lazy expiration
+  IndexSpec *spec = sctx->spec;
+  if (spec->monitorFieldExpiration) {
```

Comment:
Why not have here (and in all other such conditions) a `__builtin_expect__`? This is disabled only via our debug command no?

Reply:
I don't see why..
The indexing path is in the hot path.
How is it an overkill if the only scenario in which it is false is with an explicit DEBUG COMMAND?